buffer/src/main/java/io/netty5/buffer/adapt/AdaptablePoolingAllocator.java

```diff
@@ -0,0 +1,420 @@
+ * under the License.
+ */
+package io.netty5.buffer.adapt;
+
+import io.netty5.buffer.AllocationType;
+import io.netty5.buffer.AllocatorControl;
+import io.netty5.buffer.Buffer;
+import io.netty5.buffer.BufferAllocator;
+import io.netty5.buffer.Drop;
+import io.netty5.buffer.MemoryManager;
+import io.netty5.buffer.StandardAllocationTypes;
+import io.netty5.buffer.internal.ArcDrop;
+import io.netty5.buffer.internal.CleanerDrop;
+import io.netty5.buffer.internal.InternalBufferUtils;
+import io.netty5.util.NettyRuntime;
+
+import java.lang.invoke.MethodHandles;
+import java.lang.invoke.VarHandle;
+import java.util.Arrays;
+import java.util.concurrent.ConcurrentLinkedQueue;
+import java.util.concurrent.locks.StampedLock;
+import java.util.function.Supplier;
+
+import static io.netty5.buffer.internal.InternalBufferUtils.allocatorClosedException;
+import static io.netty5.buffer.internal.InternalBufferUtils.standardDrop;
+import static io.netty5.util.internal.PlatformDependent.threadId;
+
+public class AdaptablePoolingAllocator implements BufferAllocator {
+    private static final int RETIRE_CAPACITY = 4 * 1024;
+    private static final int MIN_CHUNK_SIZE = 128 * 1024;
+    private static final int MAX_STRIPES = NettyRuntime.availableProcessors() * 2;
+
+    private final AllocationType allocationType;
+    private final MemoryManager manager;
+    private final ConcurrentLinkedQueue<Buffer> centralQueue;
+    private final AllocatorControl allocatorControl;
+    private final StampedLock magazineExpandLock;
+    private volatile Magazine[] magazines;
+    private volatile boolean closed;
+
+    public AdaptablePoolingAllocator(boolean direct) {
+        this(MemoryManager.instance(), direct);
+    }
+
+    public AdaptablePoolingAllocator(MemoryManager manager, boolean direct) {
+        allocationType = direct ? StandardAllocationTypes.OFF_HEAP : StandardAllocationTypes.ON_HEAP;
+        this.manager = manager;
+        centralQueue = new ConcurrentLinkedQueue<>();
+        allocatorControl = new SimpleAllocatorControl(this);
+        magazineExpandLock = new StampedLock();
+        Magazine[] mags = new Magazine[4];
+        for (int i = 0; i < mags.length; i++) {
+            mags[i] = new Magazine(this);
+        }
+        magazines = mags;
+    }
+
+    @Override
+    public boolean isPooling() {
+        return true;
+    }
+
+    @Override
+    public AllocationType getAllocationType() {
+        return allocationType;
+    }
+
+    @Override
+    public Buffer allocate(int size) {
+        if (closed) {
+            throw allocatorClosedException();
+        }
+        InternalBufferUtils.assertValidBufferSize(size);
+        int sizeBucket = AllocationStatistics.sizeBucket(size); // Compute outside of Magazine lock for better ILP.
+        int expansions = 0;
+        do {
+            Magazine[] mags = magazines;
+            int mask = mags.length - 1;
+            int index = (int) (threadId(Thread.currentThread()) & mask);
```

Comment:
If an expansion happen, some of the locality will be lost, per thread, because masking won't make it reach anymore the previously chosen magazine, is it correct?

Reply:
Yes that's correct. Expansions should be rare, though.

Reply:
It is also intentional, since we expanded because the magazine we wanted was locked.